### PR TITLE
Enhance Push-OutputBinding by using the output binding information

### PR DESCRIPTION
**Pull request by daxian-dbw**
Enhance Push-OutputBinding by using the output binding information.

### CODE REVIEW

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 [hunk 1]
```diff
@@ -107,46 +215,96 @@ function Push-KeyValueOutputBinding {
...
-            Position=0,
-            ValueFromPipelineByPropertyName=$true)]
-        [string]
-        $Name,
-
-        [Parameter(
-            Mandatory=$true,
-            ParameterSetName="NameValue",
-            Position=1,
-            ValueFromPipelineByPropertyName=$true)]
```
Comment (daxian-dbw) [inline src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]: I removed the `ValueFromPipelineByPropertyName` from both `-Name` and `-Value`, as well as the `-InputObject`.
My argument is that I don't think its current pipeline parameter binding behavior is useful in real scenarios. Currently it supports:
```
[pscustomobject] @{ Name = 'queue'; Value = 'xxx' } | Push-OutputBinding
@{ Name = 'queue'; Value = 'xxx' } | Push-OutputBinding
```
The proposed new behavior is:
```
<# produce values #> | Push-OutputBinding -Name queue
```

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 [hunk 2]
```diff
@@ -4,16 +4,26 @@
...
 # These variables hold the ScriptBlock and CmdletInfo objects for constructing a SteppablePipeline of 'Out-String | Write-Information'.
 $script:outStringCmd = $ExecutionContext.InvokeCommand.GetCommand("Microsoft.PowerShell.Utility\Out-String", [CommandTypes]::Cmdlet)
 $script:writeInfoCmd = $ExecutionContext.InvokeCommand.GetCommand("Microsoft.PowerShell.Utility\Write-Information", [CommandTypes]::Cmdlet)
 $script:tracingSb = { & $script:outStringCmd -Stream | & $script:writeInfoCmd -Tags "__PipelineObject__" }
 # This loads the resource strings.
 Import-LocalizedData LocalizedData -FileName PowerShellWorker.Resource.psd1
 
+# Enum that defines different behaviors when collecting output data
+enum DataCollectingBehavior {
+    Scalar
```
Comment (TylerLeonhardt) [inline src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]: Do you consider a single object (i.e. hashtable) as a `Scalar`? I feel like `Singleton` might be better here. Not blocking on this though.

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 [hunk 3]
```diff
@@ -67,29 +77,127 @@ function Get-OutputBinding {
...
+        {
+            $list.Add($item)
+        }
+    }
+    else
+    {
+        $list.Add($NewData)
+    }
+
+    return ,$list
```
Comment (TylerLeonhardt) [inline src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]: The `,` assures we always return a list, right?

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 [hunk 4]
```diff
@@ -107,46 +215,96 @@ function Push-KeyValueOutputBinding {
...
+
+    Begin
+    {
+        $bindingInfo = Get-BindingInfo -Name $Name
+        $behavior = Get-DataCollectingBehavior -BindingInfo $bindingInfo
+    }
+
+    process
+    {
+        $bindingType = "UnKnown"
```
Comment (TylerLeonhardt) [inline src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]:
```suggestion
        $bindingType = "Unknown"
```

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/PowerShellWorker.Resource.psd1 [hunk 5]
```diff
@@ -5,6 +5,9 @@
 
 ConvertFrom-StringData @'
 ###PSLOC
-    OutputBindingAlreadySet=Output binding '{0}' is already set. To override the value, use -Force.
+    OutputBindingAlreadySet=The output binding '{0}' is already set with a value. The tyep of the output binding is '{1}'. It only accepts one message/record/file per a Function invocation. To override the value, use -Clobber.
```
Comment (TylerLeonhardt) [inline src/Modules/Microsoft.Azure.Functions.PowerShellWorker/PowerShellWorker.Resource.psd1]:
```suggestion
    OutputBindingAlreadySet=The output binding '{0}' is already set with a value. The type of the output binding is '{1}'. It only accepts one message/record/file per a Function invocation. To override the value, use -Clobber.
```

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 (hunk 2 above)
Comment (daxian-dbw) [reply src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]: Yes, I also believe `Singleton` is better. Will change.

File: src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1 (hunk 3 above)
Comment (daxian-dbw) [reply src/Modules/Microsoft.Azure.Functions.PowerShellWorker/Microsoft.Azure.Functions.PowerShellWorker.psm1]: Yes, when getting here, we want to return the `List<object>` object and use it as the value, so we can reuse the list for the next `Push-OutputBinding`.
Without the comma, the list will be unrolled and turned to an array, and that will force us to create another list for the next `Push-OutputBinding`.

File: test/Unit/Modules/Microsoft.Azure.Functions.PowerShellWorker.Tests.ps1 [hunk 6]
```diff
@@ -6,6 +6,45 @@
 Describe 'Azure Functions PowerShell Langauge Worker Helper Module Tests' {
 
     BeforeAll {
+
+        ## Create the type 'FunctionMetadata' for our test
+        $code = @'
```
Comment (TylerLeonhardt) [inline test/Unit/Modules/Microsoft.Azure.Functions.PowerShellWorker.Tests.ps1]: Do PowerShell classes work here or no?

Comment (daxian-dbw) [reply test/Unit/Modules/Microsoft.Azure.Functions.PowerShellWorker.Tests.ps1]: It should work too :)

**Pull request merged**